service/service_setup.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package service
+
+import (
+	"time"
+
+	"go.uber.org/fx/config"
+	"go.uber.org/fx/metrics"
+	"go.uber.org/fx/tracing"
+	"go.uber.org/fx/ulog"
+
+	"github.com/go-validator/validator"
+	"github.com/pkg/errors"
+)
+
+func (svc *serviceCore) setupLogging(cfg config.Provider) {
+	if svc.log == nil {
+		logBuilder := ulog.Builder()
+		// load and configure logging
+		err := svc.configProvider.Get("logging").PopulateStruct(&svc.logConfig)
+		if err != nil {
+			ulog.Logger().Info("Logging configuration is not provided, setting to default logger", "error", err)
+		}
+		svc.log = logBuilder.WithConfiguration(svc.logConfig).Build()
+	} else {
+		svc.log.Debug("Using custom log provider due to service.WithLogger option")
+	}
+}
+
+func (svc *serviceCore) setupStandardConfig(cfg config.Provider) error {
+	if err := cfg.Get("").PopulateStruct(&svc.standardConfig); err != nil {
+		return errors.Wrap(err, "unable to load standard configuration")
+	}
+
+	if errs := validator.Validate(svc.standardConfig); errs != nil {
+		svc.Logger().Error("Invalid service configuration", "error", errs)
+		return errors.Wrap(errs, "service configuration failed validation")
+	}
+	return nil
+}
+
+func (svc *serviceCore) setupMetrics() {
+	if svc.Metrics() == nil {
+		svc.metrics, svc.statsReporter, svc.metricsCloser = metrics.RootScope(svc)
+		metrics.Freeze()
+	}
+}
+
+func (svc *serviceCore) setupRuntimeMetricsCollector(cfg config.Provider) error {
+	if svc.RuntimeMetricsCollector() == nil {
```

Comment:
nit: prefer early returns:

```go
if svc.RuntimeMetricsColector() != nil {
  return nil
}

// logic to set up collector
```

this is more idiomatic and reduced indentation and increases readability.